### PR TITLE
Add affine AutoEP checkpoint placement

### PR DESCRIPTION
**Pull request by jinyouzhi**
Follow up #8385 

This pull request introduces support for flexible expert placement in DeepSpeed's AutoEP (Automatic Expert Placement) system, enabling non-uniform, non-contiguous, and replicated expert layouts. The main changes add a versioned expert placement descriptor, validation logic, and integration into checkpoint consolidation and metadata validation. This lays the groundwork for more advanced expert scheduling and model parallelism strategies.

The most important changes are:

**AutoEP Expert Placement Descriptor and Affine Map Lowering**
- Added a new module `autoep_affine.py` that defines the expert placement descriptor, validation, legacy uniform descriptor synthesis, and lowering to affine maps for sharded tensor reconstruction. This enables flexible, versioned expert placement beyond the legacy uniform contiguous layout.

**Integration into Checkpoint Consolidation and Metadata**
- Updated `autoep_universal.py` to:
  - Accept and validate the new expert placement descriptor in layer metadata, relaxing the requirement that `num_local_experts * ep_size == num_experts` when a placement is provided.
  - Use the placement descriptor and affine map for reconstructing full expert tensors during checkpoint consolidation, supporting arbitrary expert layouts. [[1]](diffhunk://#diff-ee96c370ed05cdc2757c212d5a407fbc3bd4370814512fbfa9041a8e2189a179L291-R345) [[2]](diffhunk://#diff-ee96c370ed05cdc2757c212d5a407fbc3bd4370814512fbfa9041a8e2189a179R360-R377)
  - Validate placement consistency during expert file consolidation.
  - Import and use the new placement logic. [[1]](diffhunk://#diff-ee96c370ed05cdc2757c212d5a407fbc3bd4370814512fbfa9041a8e2189a179R19-R27) [[2]](diffhunk://#diff-ee96c370ed05cdc2757c212d5a407fbc3bd4370814512fbfa9041a8e2189a179R49)

**Metadata Validation Enhancements**
- Updated `autoep_zero3_metadata.py` to:
  - Import and use placement validation and legacy descriptor synthesis.
  - Track and validate placements for each layer entry, and check runtime layer consistency. [[1]](diffhunk://#diff-cd9bddd85f897f40b30162d7fb86518be97fd1868327635349cb7b56393ed6f2R50) [[2]](diffhunk://#diff-cd9bddd85f897f40b30162d7fb86518be97fd1868327635349cb7b56393ed6f2R59) [[3]](diffhunk://#diff-cd9bddd85f897f40b30162d7fb86518be97fd1868327635349cb7b56393ed6f2R79-R82)
  - Handle the presence or absence of the placement descriptor during partitioned metadata validation.

**Documentation Updates**
- Expanded the affine IR specification (`affine_ir_spec.md`) to document the new AutoEP placement descriptor, its semantics, and its integration into the IR and runtime, clarifying the distinction between placement provenance and scheduling.

**Bugfixes and Robustness**
- Fixed a potential bug in `affine.py` by skipping empty piece lists during tensor rebuilding, preventing errors when a rank has no assigned pieces.

Related: https://github.com/deepspeedai/DeepSpeed/issues/8252, https://github.com/deepspeedai/DeepSpeed/issues/8230.

### CODE REVIEW

**Comment by delock** [comment]
Hi @jinyouzhi thank you for your PR.  This PR overall looks good to me, I have two questions:
1. did you tested that the new checkpoint format can be correctly converted to UC during training?
2. I see that UC info in checkpoint contains dict rather than affine map.   Lower to affine map happens during convertion to UC checkpoint.   For AutoTP, affine map is written in UC info in checkpoint.  Is this difference a design choice?

**Comment by jinyouzhi** [comment]
Thank you for comments.
For 1, I am design e2e test to validate loss curve consistency ( 50 steps -> save/load -> 50 steps vs 100 steps)
For 2, aggree with your observation, AutoEP should align with AutoTP by persisting a versioned per-parameter affine map at save time. I will retain the placement descriptor for provenance and legacy fallback.
@delock

**Comment by jinyouzhi** [comment]
@delock 

I completed an end-to-end loss-resume validation using the DeepSpeedExamples script (https://github.com/deepspeedai/DeepSpeedExamples/pull/1014 ) `training/deepspeed_finetune_demo/run_autoep_affine_ir_checkpoint_experiment.sh` , with Moonlight-16B-A3B on 4 GPUs (AutoEP=4, ZeRO-3 with CPU parameter/optimizer offload).
 
 The experiment trains an uninterrupted 100-step baseline, saves a native checkpoint at step 50, then resumes both directly from the native checkpoint and after  ds_to_universal.py  conversion. The converter completed the AutoEP ZeRO-3 expert-state consolidation successfully, and all runs had finite loss values. For every resumed step (51–100), the native-resume and Universal-checkpoint-resume losses are exactly identical. Their shared difference from the uninterrupted baseline is  max_abs_error=0.1374  and  mean_abs_error=0.01052 ; because the native and Universal paths match at every step, this difference is not introduced by the Universal/AffineIR conversion.
 
 I attached the loss plot below.
 
loss_comparison

File: deepspeed/checkpoint/autoep_universal.py [hunk 1]
```diff
@@ -288,19 +336,27 @@ def consolidate_autoep_zero12_expert_states(temp_dir, output_dir, expert_param_i
```
Comment (delock) [inline deepspeed/checkpoint/autoep_universal.py:336]: is `consolidate_autoep_zero12_expert_states` still needed?

Comment (delock) [reply deepspeed/checkpoint/autoep_universal.py:336]: also if it is no longer called, does it mean this functionality is lost?  Is it intentionally not called?

**Comment by jinyouzhi** [comment]
fix ZeRO1/2 issues and also have a test with ZeRO-2 @delock 

loss_zero2

File: deepspeed/checkpoint/autoep_universal.py (hunk 1 above)
Comment (jinyouzhi) [reply deepspeed/checkpoint/autoep_universal.py:304]: fixed. thanks

**Review by delock — APPROVED** [review]
